interfaces/builtin/custom_device_test.go

```diff
@@ -0,0 +1,493 @@
+}
+
+func (s *CustomDeviceInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "custom-device")
+}
+
+func (s *CustomDeviceInterfaceSuite) TestSanitizePlug(c *C) {
+	c.Check(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+	c.Check(interfaces.BeforeConnectPlug(s.iface, s.plug), IsNil)
+}
+
+func (s *CustomDeviceInterfaceSuite) TestSanitizePlugUnhappy(c *C) {
+	var customDeviceYaml = `name: consumer
+version: 0
+plugs:
+ hwdev:
+  interface: custom-device
+  %s
+apps:
+ app:
+  plugs: [hwdev]
+`
+	data := []struct {
+		plugYaml      string
+		expectedError string
+	}{
+		{
+			"custom-device: [one two]",
+			`custom-device "custom-device" attribute must be a string, not \[one two\]`,
+		},
+	}
+
+	for _, testData := range data {
+		snapYaml := fmt.Sprintf(customDeviceYaml, testData.plugYaml)
+		_, plug := MockConnectedPlug(c, snapYaml, nil, "hwdev")
+		err := interfaces.BeforePreparePlug(s.iface, plug)
+		c.Check(err, ErrorMatches, testData.expectedError, Commentf("yaml: %s", testData.plugYaml))
+	}
+}
+
+func (s *CustomDeviceInterfaceSuite) TestPlugNameAttribute(c *C) {
+	var plugYamlTemplate = `name: consumer
+version: 0
+plugs:
+ hwdev:
+  interface: custom-device
+  %s
+apps:
+ app:
+  plugs: [hwdev]
+`
+
+	data := []struct {
+		plugYaml     string
+		expectedName string
+	}{
+		{
+			"",      // missing "custom-device" attribute
+			"hwdev", // use the name of the plug
+		},
+		{
+			"custom-device: shmemFoo",
+			"shmemFoo",
+		},
+	}
+
+	for _, testData := range data {
+		snapYaml := fmt.Sprintf(plugYamlTemplate, testData.plugYaml)
+		_, plug := MockConnectedPlug(c, snapYaml, nil, "hwdev")
+		err := interfaces.BeforePreparePlug(s.iface, plug)
+		c.Assert(err, IsNil)
+		c.Check(plug.Attrs["custom-device"], Equals, testData.expectedName,
+			Commentf(`yaml: %q`, testData.plugYaml))
+	}
+}
+
+func (s *CustomDeviceInterfaceSuite) TestSanitizeSlot(c *C) {
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.slotInfo), IsNil)
+}
+
+func (s *CustomDeviceInterfaceSuite) TestSanitizeSlotUnhappy(c *C) {
+	var customDeviceYaml = `name: provider
+version: 0
+slots:
+ hwdev:
+  interface: custom-device
+  %s
+apps:
+ app:
+  slots: [hwdev]
+`
+	data := []struct {
+		slotYaml      string
+		expectedError string
+	}{
+		{
+			"custom-device: [one two]",
+			`custom-device "custom-device" attribute must be a string, not \[one two\]`,
+		},
+		{
+			"devices: 12",
+			`snap "provider" has interface "custom-device" with invalid value type int64 for "devices" attribute.*`,
+		},
+		{
+			"read-devices: [/dev/zero, 2]",
+			`snap "provider" has interface "custom-device" with invalid value type \[\]interface {} for "read-devices" attribute.*`,
+		},
+		{
+			"devices: [/dev/foo**]",
+			`custom-device "devices" path contains invalid glob pattern "\*\*"`,
+		},
+		{
+			"devices: [/dev/@foo]",
+			`custom-device "devices" path must start with / and cannot contain special characters.*`,
+		},
+		{
+			"devices: [/dev/foo|bar]",
```

Comment:
AFAICT we are not testing the also excluded chars `"{}` here, maybe we could do a tiny followup with an extra testline here for those?

Reply:
I pushed one commit more :-)